yt/data_objects/data_containers.py

```diff
@@ -427,20 +401,73 @@ def _parameter_iterate(self, seq):
 
     _key_fields = None
     def write_out(self, filename, fields=None, format="%0.16e"):
-        if fields is None: fields=sorted(self.field_data.keys())
-        if self._key_fields is None: raise ValueError
-        field_order = self._key_fields[:]
+        """Write out the YTDataContainer object in a text file.
+
+        This function will take a data object and produce a tab delimited text
+        file containing the fields presently existing and the fields given in
+        the ``fields`` list.
+
+        Parameters
+        ----------
+        filename : String
+            The name of the file to write to.
+
+        fields : List of string, Default = None
+            If this is supplied, these fields will be added to the list of
+            fields to be saved to disk. If not supplied, whatever fields
+            presently exist will be used.
+
+        format : String, Default = "%0.16e"
+            Format of numbers to be written in the file.
+
+        Raises
+        ------
+        ValueError
+            Raised when there is no existing field.
+
+        YTException
+            Raised when field_type of supplied fields is inconsistent with the
+            field_type of existing fields.
+
+        Examples
+        --------
+        >>> ds = fake_particle_ds()
+        >>> sp = ds.sphere(ds.domain_center, 0.25)
+        >>> sp.write_out("sphere_1.txt")
+        >>> sp.write_out("sphere_2.txt", fields=["cell_volume"])
+        """
+        if fields is None:
+            fields = sorted(self.field_data.keys())
+
+        if self._key_fields is None:
+            raise ValueError
+
+        field_order = self._key_fields
+        diff_fields = [field for field in fields if field not in field_order]
+        field_order += diff_fields
+        field_order = sorted(self._determine_fields(field_order))
+        field_types = {u for u, v in field_order}
+
+        if len(field_types) != 1:
+            diff_fields = self._determine_fields(diff_fields)
+            req_ftype = self._determine_fields(self._key_fields[0])[0][0]
+            f_type = {f for f in diff_fields if f[0] != req_ftype }
+            msg = ("Field type %s of the supplied field %s is inconsistent"
+                   " with field type '%s'." %
+                   ([f[0] for f in f_type], [f[1] for f in f_type], req_ftype))
+            raise YTException(msg)
+
         for field in field_order: self[field]
-        field_order += [field for field in fields if field not in field_order]
-        fid = open(filename,"w")
-        fid.write("\t".join(["#"] + field_order + ["\n"]))
+        fid = open(filename, "w")
```

Comment:
I'd be happier with a `with` statement there, to prevent the file handler from being left opened.

Reply:
done